File: haskell/space-age/SpaceAge.hs
module SpaceAge(Planet(..), ageOn) where

data Planet = Mercury | Venus | Earth | Mars | Jupiter | Saturn | Uranus | Neptune

ageOn :: (Fractional c, Num c) => Planet -> c -> c
ageOn Mercury seconds = seconds / 7600543.81992
ageOn Venus   seconds = seconds / 19414149.052176
ageOn Earth   seconds = seconds / 31557600
ageOn Mars    seconds = seconds / 59354032.69008
ageOn Jupiter seconds = seconds / 374355659.124
ageOn Saturn  seconds = seconds / 929292362.8848
ageOn Uranus  seconds = seconds / 2651370019.3296
ageOn Neptune seconds = seconds / 5200418560.0320005

File: lua/space-age/space-age.lua
local space_age    = {}
local space_age_mt = {}

space_age.new = function(_, seconds)
    local self = { seconds = seconds }
    return setmetatable(self, space_age_mt)
end

local year_duration_at = {
    earth   = 31557600,
    mercury = 31557600 * 0.2408467,
    venus   = 31557600 * 0.61519726,
    mars    = 31557600 * 1.8808158,
    jupiter = 31557600 * 11.862615,
    saturn  = 31557600 * 29.447498,
    uranus  = 31557600 * 84.016846,
    neptune = 31557600 * 164.79132,
}

local function round(number, decimals)
  local factor = 10^(decimals or 0)
  return math.floor(number * factor + 0.5) / factor
end

local age_on = function(self, planet)
    return round(self.seconds / year_duration_at[planet], 2)
end

space_age_mt.__index = function(object, key)
    local found, _, planet = key:find('^on_(%w+)')
    if found and year_duration_at[planet] then
        -- I would create a table to validate 'planet', but this one works.
        return function() return age_on(object, planet) end
    end

    return space_age[key]
end

return space_age
